### DISCUSSION TITLE
Rancher Desktop v1.13.1 - Port Forwarding not binding to ipv4 in MacOS

### DISCUSSION THREAD

**Discussion by dcardozo** [rancher-sandbox/rancher-desktop#6926] [Q&A]
I'm using Rancher Desktop v1.13.1 in a MacBook Pro (m1) running MacOS 14.3.1. I've deployed a Java application in kubernetes and enabled port forwarding via the Rancher Desktop UI.

Trying to connect to the service from a Java client application in the localhost is met with a connection refused message; if I disable port forwarding via UI and enable it via kubectl, it works fine. The output of netstat shows that forwarding via the UI binds the port to ipv6 only, while forwarding via kubectl binds the port to both ipv4 and ipv6.

In version 1.11.1, port forwarding via the UI bound the port to ipv4 only.

### Questions:
1. Is the change in behavior intentional? i.e. Rancher Desktop 1.11.1 binding to ipv4 only and 1.13.1 binding to ipv6 only.
2. Is there any available setting in Rancher Desktop that would allow binding ports to both ipv4 and ipv6 (same as kubectl port-forward)?
3. Alternatively, is there a setting to allow reverting back to version 1.11.1 behavior?

Thank you in advance for any guidance you may provide,

David

---
I've provided a simple Java Socket program to illustrate the issue:
#### SimpleServer.java:
```java
import java.io.OutputStream;
import java.net.ServerSocket;
import java.net.Socket;

public class SimpleServer {
    public static void main(String[] args) {
        int port = 8000;

        try (ServerSocket serverSocket = new ServerSocket(port)) {
            System.out.println("Server is listening on port " + port);

            while (true) {
                try (Socket socket = serverSocket.accept()) {
                    System.out.println("New client connected");

                    OutputStream outputStream = socket.getOutputStream();
                    outputStream.write("success".getBytes());
                    outputStream.flush();
                } catch (Exception e) {
                    System.err.println("Error handling client connection: " + e.getMessage());
                }
            }
        } catch (Exception e) {
            System.err.println("Server error: " + e.getMessage());
        }
    }
}
```
#### SimpleClient.java
```java
import java.io.InputStream;
import java.net.Socket;

public class SimpleClient {
    public static void main(String[] args) {
        String host = "localhost";
        int port = 8000;

        try (Socket socket = new Socket(host, port)) {
            InputStream inputStream = socket.getInputStream();
            byte[] buffer = new byte[7]; // "success" is 7 bytes long
            int read = inputStream.read(buffer);
            String message = new String(buffer, 0, read);

            System.out.println("Received message: " + message);
        } catch (Exception e) {
            System.err.println("Client error: " + e.getMessage());
        }
    }
}
```
#### Dockerfile
```
# Stage 1: Build the Java application
FROM openjdk:11-jdk-slim AS build

# Set the working directory inside the build container
WORKDIR /app

# Copy the Java source code to the build container
COPY SimpleServer.java /app

# Compile the Java source code
RUN javac SimpleServer.java

# Stage 2: Create the final runtime image
FROM openjdk:11-jre-slim

# Set the working directory inside the runtime container
WORKDIR /app

# Copy the compiled class file from the build container
COPY --from=build /app/SimpleServer.class /app

# Expose port 8000 for the application
EXPOSE 8000

# Run the SimpleServer application
CMD ["java", "SimpleServer"]
```
#### simple-server.yml
```yml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: simple-server-deployment
spec:
  replicas: 1
  selector:
    matchLabels:
      app: simple-server
  template:
    metadata:
      labels:
        app: simple-server
    spec:
      containers:
        - name: simple-server
          image: test/simple-server:1.0.0
          ports:
            - containerPort: 8000
---
apiVersion: v1
kind: Service
metadata:
  name: simple-server-service
spec:
  selector:
    app: simple-server
  ports:
    - protocol: TCP
      port: 8000
      targetPort: 8000
  type: ClusterIP

```

Here are the steps to reproduce and the output:

Build image: ` docker build -t test/simple-server:1.0.0 .`
Compile client program: ` javac SimpleClient.java`

### Test 1
Run as a simple  container: `docker run --rm -p 8000:8000 test/simple-server:1.0.0`
Run client: `java SimpleClient`
Output: `Received message: success`
netstat output:
```
❯ netstat -an | grep 8000
tcp4       0      0  *.8000                 *.*                    LISTEN
tcp4       0      0  127.0.0.1.8000         127.0.0.1.49605        TIME_WAIT
```
### Test 2
Deploy in kubernetes: `kubectl apply -f simple-server.yml`
Enable port forwarding via Rancher Desktop UI
Run client: `java SimpleClient`
Output: `Client error: Connection refused`
netstat output:
```
❯ netstat -an | grep 8000
tcp6       0      0  ::1.8000               *.*                    LISTEN
```

### Test 3
Stop port forwarding in the UI, enable port forwarding via kubectl: `kubectl port-forward services/simple-server-service 8000:8000`
kubectl output:
```
Forwarding from 127.0.0.1:8000 -> 8000
Forwarding from [::1]:8000 -> 8000
```

Run client: `java SimpleClient`
Output: `Received message: success`
netstat output:
```
❯ netstat -an | grep 8000
tcp6       0      0  ::1.8000               *.*                    LISTEN
tcp4       0      0  127.0.0.1.8000         *.*                    LISTEN
tcp4       0      0  127.0.0.1.49723        127.0.0.1.8000         TIME_WAIT
```

---
Downgrading to Rancher Desktop v1.11.1 allows me to enable port forwarding via the UI and connect successfully:

### Test 1
Same behavior as in Rancher Desktop v1.13.1

### Test 2
Deploy in kubernetes: `kubectl apply -f simple-server.yml`
Enable port forwarding via Rancher Desktop UI
Run client: `java SimpleClient`
Output: `Received message: success`
netstat output:
```
❯ netstat -an | grep 8000
tcp4       0      0  127.0.0.1.8000         *.*                    LISTEN
tcp4       0      0  127.0.0.1.49350        127.0.0.1.8000         TIME_WAIT
```

### Test 3
Same behavior as in Rancher Desktop v1.13.1

**Comment by uromahn** [comment]
This is not an answer but a confirmation that the same happens on Rancher Desktop 1.15.1 running on Mac OS Sonoma 14.6.1.
This is super annoying, especially since it worked before on RD 1.14!
I really don't want to downgrade but it appears I have no choice doing that for now.

**Comment by e40** [comment]
I'm seeing similar behavior.  I have two containers, one I want to be available on my network and the other which I do not.  The former works fine.  The latter does not, giving me "Connection refused".

Running RD 1.16.0.  k8s is off.

On macOS 14.7.

```
@max[git:master]$ wget http://localhost:9091
--2024-09-29 10:21:46--  http://localhost:9091/
Resolving localhost (localhost)... ::1, 127.0.0.1
Connecting to localhost (localhost)|::1|:9091... failed: Connection refused.
Connecting to localhost (localhost)|127.0.0.1|:9091... failed: Connection refused.
```

`docker ps` shows `127.0.0.1:9091->9091/tcp`.  And I can verify the guest side:

```
@max[git:master]$ docker exec readme netstat -an
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       
tcp        0      0 0.0.0.0:9090            0.0.0.0:*               LISTEN      
Active UNIX domain sockets (servers and established)
Proto RefCnt Flags       Type       State         I-Node Path
@max[git:master]$ 
```

The exact same scripts work on Linux.  Seems like a macOS bug with RD.
